### PR TITLE
Refactor(eos_designs): Prevent empty path-groups for auto generated WAN policies

### PR DESCRIPTION
**Pull request by gmuloc**
## Change Summary

Current implementation could potentially allow empty (as in no path-group) load balancing policies for AVD auto generated policies (also control-plane) if all path-groups have `excluded_from_default_policy` set to True

This PR checks that this is not the case and that for a given device the list would not be empty either (this could be triggered on some inventories and not others)

## Component(s) name

`arista.avd.eos_designs`

## How to test
molecule negative unit tests

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 1]
```diff
@@ -312,6 +299,49 @@ def _filtered_wan_policies(self) -> list:
...
+        """
+        path_groups = [
+            path_group["name"] for path_group in self.shared_utils.wan_path_groups if not get(path_group, "excluded_from_default_policy", default=False)
+        ]
+        local_path_group_names = [path_group["name"] for path_group in self.shared_utils.wan_local_path_groups]
+        if not path_groups:
+            # all path-groups are excluded from default-policy
+            raise AristaAvdError(
+                "Empty list of path-groups to use for the AVD autogenerated default policy, either `wan_path_groups` is empty "
+                "or all path-groups have `excluded_from_default_policy` set tot `true`. Make sure that at least one path-group can be "
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py:322]:
```suggestion
                "or all path-groups have `excluded_from_default_policy` set to `true`. Make sure that at least one path-group can be "
```

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 2]
```diff
@@ -312,6 +299,49 @@ def _filtered_wan_policies(self) -> list:
...
+                "configured for the default-policy"
+            )
+        elif not [path_group for path_group in path_groups if path_group in local_path_group_names]:
+            # No common path-group between this device local path-groups and the available path-group for the default policy
+            raise AristaAvdError(
+                f"None of this device local path-groups {local_path_group_names} is eligible to be included in the "
+                "AVD default-policy. Make sure that at least one path-group for the device is not configured with "
+                "`excluded_from_default_policy` set tot `true` under `wan_path_groups`."
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]:
```suggestion
                "`excluded_from_default_policy: true` under `wan_path_groups`."
```

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py (hunk 1 above)
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]:
```suggestion
                "or all path-groups have `excluded_from_default_policy` set to `true`. Make sure that at least one path-group can be "
```

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 3]
```diff
@@ -312,6 +299,49 @@ def _filtered_wan_policies(self) -> list:
...
 
+    @cached_property
+    def _default_wan_policy_name(self) -> str:
+        """
+        TODO make this configurable
+        """
+        return "DEFAULT-POLICY"
+
+    @cached_property
+    def _default_policy_path_groups(self) -> list:
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]:
```suggestion
    def _default_policy_path_group_names(self) -> list:
```

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 4]
```diff
@@ -312,6 +299,49 @@ def _filtered_wan_policies(self) -> list:
...
+        """
+        Return the list of path-groups to consider when generating a default policy with AVD
+        """
+        path_groups = [
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]:
```suggestion
        path_group_names = [
```

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 5]
```diff
@@ -312,6 +299,49 @@ def _filtered_wan_policies(self) -> list:
...
+            path_group["name"] for path_group in self.shared_utils.wan_path_groups if not get(path_group, "excluded_from_default_policy", default=False)
+        ]
+        local_path_group_names = [path_group["name"] for path_group in self.shared_utils.wan_local_path_groups]
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]: we can move this line after if statement.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 6]
```diff
@@ -312,6 +299,49 @@ def _filtered_wan_policies(self) -> list:
...
+        if not path_groups:
+            # all path-groups are excluded from default-policy
+            raise AristaAvdError(
+                "Empty list of path-groups to use for the AVD autogenerated default policy, either `wan_path_groups` is empty "
+                "or all path-groups have `excluded_from_default_policy` set tot `true`. Make sure that at least one path-group can be "
+                "configured for the default-policy"
+            )
+        elif not [path_group for path_group in path_groups if path_group in local_path_group_names]:
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]: local_path_group_names.intersection(path_groups) can help in readability.

Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]: interesction is for sets

```
>>> l1 = []
>>> l1.intersection([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute 'intersection'
```

we may go in set direction

**Review by ClausHolbechArista — APPROVED** [review]

**Review by carlbuchmann — CHANGES_REQUESTED** [review]
@gmuloc in my testing I'm seeing some double single quotes `'` in the returned list.

```
 msg: 'Unable to generate the default WAN policy as none of the device local path-groups [''MPLS'', ''Internet''] is eligible to be included```

**Comment by carlbuchmann** [comment]
> @gmuloc in my testing I'm seeing some double single quotes `'` in the returned list.
> 
> ```
>  msg: 'Unable to generate the default WAN policy as none of the device local path-groups [''MPLS'', ''Internet''] is eligible to be included```
> ```

This was caused by `stdout_callback = community.general.yaml` in ansible.cfg

**Review by carlbuchmann — APPROVED** [review]
LGTM - tested in lab

**Pull request merged**